packages/web/src/components/QueryHooksProvider.tsx

```diff
@@ -0,0 +1,84 @@
+import type { DocumentNode } from 'graphql'
+
+export interface QueryHookOptions {
+  variables?: Record<string, any>
+}
+export interface OperationResult<TData = any> {
+  data?: TData | undefined
+  loading: boolean
+  error?: Error
+}
+
+export type MutationOperationResult<TData = any> = [
+  (options?: any) => Promise<TData>,
+  OperationResult<TData>
+]
+
+export interface QueryHooks {
+  mapUseQueryHook: (
+    query: DocumentNode,
+    options?: QueryHookOptions
+  ) => OperationResult
+  mapUseMutationHook: (
+    mutation: DocumentNode,
+    options?: QueryHookOptions
+  ) => MutationOperationResult
+}
+export const QueryHooksContext = React.createContext<QueryHooks>({
+  mapUseQueryHook: () => {
+    throw new Error(
+      'You must register a useQuery hook via the `QueryHooksProvider`'
+    )
+  },
+  mapUseMutationHook: () => {
+    throw new Error(
+      'You must register a useMutation hook via the `QueryHooksProvider`'
+    )
+  },
+})
+
+/**
+ * QueryHooksProvider stores a standard `useQuery` hook for Redwood
+ * that can be mapped to your GraphQL library of choice's own `useQuery`
+ * and `useMutation` implementation.
+ *
+ * @todo Let the user pass in the additional type for options.
+ */
+export const QueryHooksProvider: React.FunctionComponent<{
+  registerUseQueryHook: (
```

Comment:
Yeah - I had some trouble with these since I don't want the user to just pass in `useQuery`, but instead want them to map the parameters and the return types, but you're right, register does seem a bit weird.

Reply:
Ohh, now the naming makes more sense. And the internal `mapUseQueryHook` too!
Can we get away with naming it `useQuery`, and then relying on the typings to make the user understand that they need to pass in a function that mapps their useQuery signature to the signature RW expects?

Reply:
Ok, I'm going to go with `useQuery|Mutation`